### ISSUE TITLE
nodejs always waits for timeout

### ISSUE THREAD

**Issue by loomis** [funcool/promesa#79]
I'm using promesa within a NodeJS script to interact with a remote server via an asynchronous API.  The main function of the script essentially looks like this:

```clj
(let [promise  (p/deferred)
          callback (fn [v] (p/resolve! promise v))]
      (async-call-api callback)
      (-> promise
          (timeout* 5000)
          (p/then (fn [result] (println "result: " result "\n")))
          (p/catch #(println "command timed out"))))
```

Because `timeout` indirectly creates a timer, NodeJS will wait for the result (assuming the server responds) because it will not terminate while there are still active timers. Unfortunately, the timer is never cancelled, so NodeJS will _always_ wait for the timeout (5s here) to expire, even if the result is available sooner. 

I can work around this by changing the `timeout` implementation to:

```clj 
(defn timeout*
  "Returns a cancellable promise that will be fulfilled with this
  promise's fulfillment value or rejection reason.  However, if this
  promise is not fulfilled or rejected within `ms` milliseconds, the
  returned promise is cancelled with a TimeoutError"
  ([p t] (timeout* p t ::default exec/default-scheduler))
  ([p t v] (timeout* p t v exec/default-scheduler))
  ([p t v scheduler]
   (c/let [timeout (p/deferred)]
     (let [tid             (exec/schedule! scheduler t #(if (= v ::default)
                                                          (p/reject! timeout (js/Error. "Operation timed out."))
                                                          (p/resolve! timeout v)))
           wrapped-promise (p/then p (fn [result] (pt/-cancel! tid) result))]
       (p/race [wrapped-promise timeout])))))
```

This explicitly cancels the timeout task if the given promise returns. Is there a downside to cancelling the timeout task explicitly? Should this be the default behavior for `timeout`? I don't know if this would have negative effects when used from a browser or from Clojure.

**Comment by loomis** [comment]
This is the implementation that I'm using: 

```clj 
(defn timeout
  "Returns a cancellable promise that will be fulfilled with this
  promise's fulfillment value or rejection reason.  However, if this
  promise is not fulfilled or rejected within `ms` milliseconds, the
  returned promise is cancelled with a TimeoutError"
  ([p t] (timeout p t ::default exec/default-scheduler))
  ([p t v] (timeout p t v exec/default-scheduler))
  ([p t v scheduler]
   (c/let [timeout (p/deferred)
           tid     (exec/schedule! scheduler t #(if (= v ::default)
                                                  (p/reject! timeout (#?(:cljs p/TimeoutException.
                                                                         :clj  TimeoutException.)
                                                                       "Operation timed out."))
                                                  (p/resolve! timeout v)))]
     (p/race [(p/finally p #(pt/-cancel! tid))
              timeout]))))
```

Let me know if you'd like me to turn this into a PR.

**Comment by niwinz** [comment]
Hi @loomis
I know that this reply is too late, so with your permission I will fix it using your impl.

Thanks for reporting.

**Comment by niwinz** [comment]
Will be available on 6.0.2
